internal/clock/fake.go

```diff
@@ -0,0 +1,198 @@
+		if fc.now.Before(t.when) {
+			fc.now = t.when
+		}
+		fc.Unlock()
+		t.tick()
+		fc.Lock()
+	}
+
+	if fc.now.Before(end) {
+		fc.now = end
+	}
+	fc.Unlock()
+
+	nap()
+}
+
+// FakeTimer produces a timer that will emit a time some duration after now,
+// exposing the fake timer internals and type.
+func (fc *FakeClock) FakeTimer(d time.Duration) *FakeTimer {
+	fc.Lock()
+	defer fc.Unlock()
+
+	c := make(chan time.Time, 1)
+	t := &FakeTimer{
+		c:     c,
+		clock: fc,
+		when:  fc.now.Add(d),
+	}
+	fc.addTimer(t)
+	return t
+}
+
+// Timer produces a timer that will emit a time some duration after now.
+func (fc *FakeClock) Timer(d time.Duration) Timer {
+	return fc.FakeTimer(d)
+}
+
+func (fc *FakeClock) addTimer(t *FakeTimer) {
+	heap.Push(&fc.timers, t)
+}
+
+// After produces a channel that will emit the time after a duration passes.
+func (fc *FakeClock) After(d time.Duration) <-chan time.Time {
+	return fc.Timer(d).C()
+}
+
+// FakeAfterFunc waits for the duration to elapse and then executes a function.
+// A Timer is returned that can be stopped.
+func (fc *FakeClock) FakeAfterFunc(d time.Duration, f func()) *FakeTimer {
+	t := fc.FakeTimer(d)
+	go func() {
+		<-t.c
+		f()
+	}()
+	nap()
+	return t
+}
+
+// AfterFunc waits for the duration to elapse and then executes a function.
+// A Timer is returned that can be stopped.
+func (fc *FakeClock) AfterFunc(d time.Duration, f func()) Timer {
+	return fc.FakeAfterFunc(d, f)
+}
+
+// Now returns the current time on the fake clock.
+func (fc *FakeClock) Now() time.Time {
+	fc.Lock()
+	defer fc.Unlock()
+	return fc.now
+}
+
+// Sleep pauses the goroutine for the given duration on the fake clock.
+// The clock must be moved forward in a separate goroutine.
+func (fc *FakeClock) Sleep(d time.Duration) {
+	<-fc.After(d)
+}
+
+// FakeTimer represents a single event.
+type FakeTimer struct {
+	c     chan time.Time
+	when  time.Time
+	clock *FakeClock
+	index int
+}
+
+// C returns a channel that will send the time when it fires.
+func (t *FakeTimer) C() <-chan time.Time {
+	return t.c
+}
+
+// tick advances the clock to this timer.
+func (t *FakeTimer) tick() {
+	select {
+	case t.c <- t.when:
+	default:
+	}
+	nap()
+}
+
+// Reset adjusts the timer's scheduled time forward from now, unless it has
+// already fired.
+func (t *FakeTimer) Reset(d time.Duration) bool {
+	t.when = t.clock.now.Add(d)
+
+	// Empty the channel if already filled.
+	select {
+	case <-t.c:
+	default:
+	}
+
+	if t.index >= 0 {
+		heap.Fix(&t.clock.timers, t.index)
+		return true
+	}
+	heap.Push(&t.clock.timers, t)
+	return false
+}
+
+// Stop removes a timer from the scheduled timers.
+func (t *FakeTimer) Stop() bool {
+	if t.index < 0 {
+		return false
+	}
+
+	// Empty the channel if already filled.
+	select {
+	case <-t.c:
+	default:
+	}
+
+	t.clock.timers.Swap(t.index, len(t.clock.timers)-1)
+	t.clock.timers.Pop()
+	heap.Fix(&t.clock.timers, t.index)
+	return true
+}
+
+func nap() {
+	runtime.Gosched()
```

Comment:
I might just directly call this instead of putting it in nap, it makes this more readable I think, when I saw nap I thought you had gone back to a sleep call.